docs/understand/programming_model.rst

```diff
@@ -66,10 +73,51 @@ memory subsystem resources.
 
   Block Diagram of a CDNA2 Graphics Compute Die.
 
+Heterogeneous Programming
+=========================
+
+The HIP programming model assumes two execution contexts. One is referred to as
+*host* while compute kernels execute on a *device*. These contexts have
+different capabilities, therefor slightly different rules apply. The *host*
+execution is defined by the C++ abstract machine, while *device* execution
+follows the :ref:`SIMT model<programming_model_simt>` of HIP. These execution contexts in
+code are signified by the ``__host__`` and ``__device__`` decorators. There are
+a few key differences between the two:
+
+* The C++ abstract machine assumes a unified memory address space, meaning that
+  one can always access any given address in memory (assuming the absence of
+  data races). HIP however introduces several memory namespaces, an address
+  from one means nothing in another. Moreover, not all address spaces are
+  accessible from all contexts.
+
+  Looking at :ref:`cdna2_gcd` and :ref:`cdna3_cu`, you can see that
+  every CU has an instance of storage backing the namespace ``__shared__``. 
+  Even if the host were to have access to these regions of
+  memory, the performance benefits of the segmented memory subsystem are
+  supported by the inability of asynchronous access from the host.
+
+* Not all C++ language features map cleanly to typical device architectures,
+  some are very expensive (meaning slow) to implement on GPU devices, therefor
+  they are forbidden in device contexts to avoid users tapping into features
+  that unexpectedly decimate their program's performance. Offload devices targeted
+  by HIP aren't general purpose devices, at least not in the sense that a CPU is.
+  HIP focuses on data parallel computations and as such caters to throughput
+  optimized architectures, such as GPUs or accelerators derived from GPU
+  architectures.
+
+* Asynchrony is at the forefront of the HIP API. Computations launched on the device
+  execute asynchronously with respect to the host, and it is the user's responsibility to
+  synchronize their data dispatch/fetch with computations on the device. 
+  
+  .. note::
+    HIP does perform implicit synchronization on occasions, more advanced than other 
+    APIs such as OpenCL or SYCL, in which the responsibility of synchronization mostly 
+    depends on the user.
+
 .. _programming_model_simt:
 
-Single instruction multiple threads
-===================================
+Single instruction multiple threads (SIMT)
+==========================================
 
 The single instruction, multiple threads (SIMT) programming model behind the
```

Comment:
SIMT has already been abbreviated at the beginning of this page.

Reply:
Fixed.